plugins/syntax/numeral.ml

```diff
@@ -128,13 +128,212 @@ let warn_deprecated_decimal =
+
+let via_abstract_error () =
+  CErrors.user_err (Pp.str "'via' and 'abstract' cannot be used together.")
+
+let locate_global_sort_inductive_or_constant sigma qid =
+  let locate_sort qid =
+    match Nametab.locate_extended qid with
+    | Globnames.TrueGlobal _ -> raise Not_found
+    | Globnames.SynDef kn ->
+    match Syntax_def.search_syntactic_definition kn with
+    | [], Notation_term.NSort r ->
+       let sigma,c = Evd.fresh_sort_in_family sigma (Glob_ops.glob_sort_family r) in
+       sigma,Constr.mkSort c
+    | _ -> raise Not_found in
+  try locate_sort qid
+  with Not_found ->
+    match Smartlocate.global_with_alias qid with
+    | GlobRef.IndRef i -> sigma, Constr.mkInd i
+    | _ -> sigma, Constr.mkConst (Smartlocate.global_constant_with_alias qid)
+
+let locate_global_constructor_inductive_or_constant qid =
+  let g = Smartlocate.global_with_alias qid in
+  match g with
+  | GlobRef.ConstructRef c -> g, Constr.mkConstruct c
+  | GlobRef.IndRef i -> g, Constr.mkInd i
+  | _ -> g, Constr.mkConst (Smartlocate.global_constant_with_alias qid)
+
+let get_type env sigma c =
+  let t = Retyping.get_type_of env sigma (EConstr.of_constr c) in
+  EConstr.to_constr sigma t
+
+let elaborate_to_post_params env sigma ty_ind params =
+  match params with
+  | [] -> [||], get_constructors ty_ind
+  | _ ->
+     let pt_refs = get_constructors ty_ind in
+     let rec prod_length c = match Constr.kind c with
+       | Constr.Prod (_, _, c) -> 1 + prod_length c
+       | _ -> 0 in
+     let l =
+       let args = List.map (fun r -> ToPostCheck r) params in
+       let nb_params = List.length args in
+       List.map (fun r ->
+           let c = match r with
+             | GlobRef.VarRef v -> Constr.mkVar v
+             | GlobRef.ConstRef c -> Constr.mkConst c
+             | GlobRef.IndRef i -> Constr.mkInd i
+             | GlobRef.ConstructRef c -> Constr.mkConstruct c in
+           let l = prod_length (get_type env sigma c) - nb_params in
+           r, r, args @ List.init l (fun _ -> ToPostCopy))
+         pt_refs in
+     [|l|], pt_refs
+
+let elaborate_to_post_via env sigma ty_ind ty_name l =
+     let ty_ind = Constr.mkInd ty_ind in
+     let sigma, ty_name =
+       locate_global_sort_inductive_or_constant sigma ty_name in
+     (* retrieve constants and constructors mappings and their type *)
+     let l =
+       let read (c, i) =
+         let lc, li = c.CAst.loc, i.CAst.loc in
+         let gc, c = locate_global_constructor_inductive_or_constant c in
+         let gi, i =
+           let i = Smartlocate.global_constructor_with_alias i in
+           GlobRef.ConstructRef i, Constr.mkConstruct i in
+         let ic = Impargs.(select_stronger_impargs (implicits_of_global gc)) in
+         lc, gc, c, get_type env sigma c, ic, li, gi, i, get_type env sigma i in
+       List.map read l in
+     let module CSet = CSet.Make (Constr) in
+     let module CMap = CMap.Make (Constr) in
```

Comment:
Why not lift these to top-level in this file?  (Or are there global ones in some other file already?)